tests/scenario/test_machine_charm/test_cos_agent_e2e.py

```diff
@@ -94,50 +103,119 @@ def __init__(self, framework: Framework):
+    #  - send-remote-write
+    #  - logging-consumer
+    #  - grafana-dashboards-provider
     peer = PeerRelation("peers")
     cos_agent1 = SubordinateRelation(
         "cos-agent",
         primary_app_name="mock-principal",
         remote_unit_data=state_out.relations[0].local_unit_data,
     )
-    state_out1 = State(relations=[cos_agent1, peer]).trigger(
-        cos_agent1.changed_event(remote_unit_id=0),
+
+    ctx1 = Context(
         charm_type=GrafanaAgentMachineCharm,
         charm_root=vroot,
     )
+    state1 = State(relations=[cos_agent1, peer])
+    state_out1 = ctx1.run(cos_agent1.changed_event(remote_unit_id=0), state=state1)
+    assert state_out1.status.unit.name == "blocked"
 
     peer_out = state_out1.relations[1]
     peer_out_data = json.loads(peer_out.local_unit_data[CosAgentPeersUnitData.KEY])
     assert peer_out_data["principal_unit_name"] == "mock-principal/0"
 
-    # step 3: gagent leader is notified that the principal has touched its relation data
+    # Step 3: gagent is related to Prometheus through "send-remote-write" relation and ends
+    # in "blocked" status since there are missing relations:
+    #  - logging-consumer
+    #  - grafana-dashboards-provider
     prometheus = Relation("send-remote-write", remote_app_name="prometheus-k8s")
     cos_agent2 = SubordinateRelation(
         "cos-agent",
         primary_app_name="mock-principal",
         remote_unit_data=state_out.relations[0].local_unit_data,
     )
-    state_out2 = State(leader=True, relations=[cos_agent2, peer_out, prometheus]).trigger(
-        peer_out.changed_event(remote_unit_id=0),
+    ctx2 = Context(
         charm_type=GrafanaAgentMachineCharm,
         charm_root=vroot,
     )
-
+    state2 = State(leader=True, relations=[cos_agent2, peer_out, prometheus])
+    state_out2 = ctx2.run(peer_out.changed_event(remote_unit_id=0), state=state2)
     prom_relation_out = state_out2.relations[2]
-
     # the prometheus lib has put some data in local app data towards cos-lite.
     assert prom_relation_out.local_app_data["alert_rules"]
+
+    # Step 4: gagent is related to Loki through "logging-consumer" relation and ends
+    # in "blocked" status since there are missing relations:
+    #  - grafana-dashboards-provider
+    loki = Relation("logging-consumer", remote_app_name="lok-k8s")
+    state3 = State(leader=True, relations=[cos_agent2, peer_out, prometheus, loki])
+    state_out3 = ctx2.run(peer_out.changed_event(remote_unit_id=0), state=state3)
+    assert state_out3.status.unit.name == "blocked"
+
+    # Step 5: gagent is related to Grafana through "grafana-dashboards-provider" relation and ends
+    # in "active" status
+    grafana = Relation("grafana-dashboards-provider", remote_app_name="grafana-k8s")
+    state4 = State(leader=True, relations=[cos_agent2, peer_out, prometheus, loki, grafana])
+    state_out4 = ctx2.run(peer_out.changed_event(remote_unit_id=0), state=state4)
+    assert state_out4.status.unit.name == "active"
```

Comment:
**fyi** you can natively compare statuses: `assert state.status == ActiveStatus('foo')`